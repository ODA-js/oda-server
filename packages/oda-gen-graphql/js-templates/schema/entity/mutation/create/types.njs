<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-create-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/create/create${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input create#{entity.name}Input {
<#- for (let field of entity.create){#>
      #{field.name}: #{field.type}
<#-}#>
<#- for (let rel of entity.relations.filter(f=>f.persistent)){
const refName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
#>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{refName}<#if(!rel.single){#>]<#}#>
<#-}#>
    }
  `,
});

<#- chunkStart(`./mutations/create/create${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type create#{entity.name}Payload {
      #{entity.payloadName}: #{entity.plural}Edge
    }
  `,
});
