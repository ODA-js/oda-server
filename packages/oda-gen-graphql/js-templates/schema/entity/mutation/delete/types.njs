<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-delete-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/delete/delete${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input delete#{entity.name}Input {
<#- for (let field of entity.unique){#>
      #{field.name}: #{field.type}
<#-}#>
    } 
  `,
});

<#- chunkStart(`./mutations/delete/delete${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type delete#{entity.name}Payload {
      deletedItemId: ID
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
