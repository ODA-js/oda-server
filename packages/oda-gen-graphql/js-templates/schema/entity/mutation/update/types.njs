<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-update-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/update/update${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input update#{entity.name}Input {
    <#- for (let field of entity.update){#>
      #{field.name}: #{field.type}
    <#-}#>
    <#- for (let rel of entity.relations.filter(f=>f.persistent)){

    const createName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `create${rel.ref.entity}Input`;
    const updateName = rel.fields.length > 0 ? `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
    #>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
<#- if(!rel.embedded){#>
      #{rel.field}Unlink: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
      #{rel.field}Create: <#if(!rel.single){#>[<#}#>#{createName}<#if(!rel.single){#>]<#}#>
<#-}#>
    <#-}#>
    }
  `,
});

<#- chunkEnd(); -#>

<#- for (let rel of entity.relations.filter(f=>f.persistent && f.fields.length > 0)){
const createName = `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input`;
const updateName = `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input`;
let eEntity = rel.ref.eEntity;
  let eRels = eEntity.relations;
#>

<#- chunkStart(`./mutations/create/${createName}.ts`); -#>
<# slot('import-embed-entity-create-rel-mutation-types', createName) #>
<# slot('use-embed-entity-create-rel-mutation-types', createName) #>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{createName} {
<#- for (let field of eEntity.create){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}`
});

<#- chunkEnd(); -#>
<#- chunkStart(`./mutations/update/${updateName}.ts`); -#>
<# slot('import-embed-entity-update-rel-mutation-types', updateName) #>
<# slot('use-embed-entity-update-rel-mutation-types', updateName) #>

import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{updateName} {
<#- for (let field of eEntity.update){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}
`});
<#- chunkEnd(); -#>

<#-}#>


<#- chunkStart(`./mutations/update/update${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type update#{entity.name}Payload {
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
