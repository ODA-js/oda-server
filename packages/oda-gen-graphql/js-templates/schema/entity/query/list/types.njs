<#@ chunks "$$$main$$$" -#>
<#@ alias 'connection-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./query/list/${entity.plural}Edge.ts`); -#>
import { Type } from '../../../../common';
import gql from 'graphql-tag';
<#- slot('import-query-list-index-slot',`${entity.plural}Edge`)-#>
<#- slot('item-query-list-index-slot',`${entity.plural}Edge`)-#>
export default new Type({
  schema: gql`
    type #{entity.plural}Edge {
      node: #{entity.name}
      cursor: String!
      # put here your additiona edge fields
    }
  `,
});

<#-chunkStart(`./query/list/${entity.plural}Connection.ts`); -#>
import { Type } from '../../../../common';
import gql from 'graphql-tag';
<#- slot('import-query-list-index-slot',`${entity.plural}Connection`)-#>
<#- slot('item-query-list-index-slot',`${entity.plural}Connection`)-#>
export default new Type({
  schema: gql`
    type #{entity.plural}Connection {
      pageInfo: PageInfo!
      edges: [#{entity.plural}Edge]
      # put here your additional connection fields
    }
  `,
});
<#- chunkEnd() -#>

<#
if(entity.connections.length > 0){
  entity.connections.forEach(connection => {
#>

<#- chunkStart(`./query/list/${connection.connectionName}Connection.ts`); -#>
import { Type } from '../../../../common';
import gql from 'graphql-tag';
<#- slot('import-query-list-index-slot',`${connection.connectionName}Connection`)-#>
<#- slot('item-query-list-index-slot',`${connection.connectionName}Connection`)-#>
export default new Type({
  schema: gql`
    type #{connection.connectionName}Connection {
      pageInfo: PageInfo!
      edges: [#{connection.connectionName}Edge]
      # put here your additional connection fields
    }
  `,
});

<#- chunkStart(`./query/list/${connection.connectionName}Edge.ts`); -#>
import { Type } from '../../../../common';
import gql from 'graphql-tag';
<#- slot('import-query-list-index-slot',`${connection.connectionName}Edge`)-#>
<#- slot('item-query-list-index-slot',`${connection.connectionName}Edge`)-#>
export default new Type({
  schema: gql`
    type #{connection.connectionName}Edge {
      node: #{connection.refType}
      cursor: String!
    <#- if(connection.fields.length > 0){#>
      #additional Edge fields
    <#
    connection.fields.forEach(f=>{#>
    <#-if(f.description){#>
      # #{f.description}
    <#}-#>
      #{f.name}#{f.argsString}: #{f.type}
    <#- });-#>
    <#}#>
      # put here your additiona edge fields
    }
  `,
});
<#- chunkEnd() -#>
<#-})
}-#>