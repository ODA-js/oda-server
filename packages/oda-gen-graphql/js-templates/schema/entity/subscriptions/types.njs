<#@ chunks "$$$main$$$" -#>
<#@ alias 'subscriptions-types'#>
<#@ context 'entity' -#>

<# const hasConnections = entity.connections.length > 0; #>
<#-chunkStart(`./subscriptions/${hasConnections ? 'Update' : ''}${entity.name}SubscriptionPayload.ts`); -#>
<#- slot('export-subscriptions-slot',`${hasConnections ? 'Update' : ''}${entity.name}SubscriptionPayload`) #>
<#- slot('import-subscriptions-slot',`${hasConnections ? 'Update' : ''}${entity.name}SubscriptionPayload`) #>

import { ModelType, Type } from '../../../common';
import gql from 'graphql-tag';

export default new Type({
  type: ModelType.type,
  schema: gql`
    <#if(hasConnections){#>
    type Update#{entity.name}SubscriptionPayload {
    <#} else {#>
    type #{entity.name}SubscriptionPayload {
    <#}#>
    <#- for (let field of entity.update){#>
      #{field.name}: #{field.type}
    <#-}#>
    }
  `,
});

<#-chunkStart(`./subscriptions/${entity.name}Subscription.ts`); -#>
<#- slot('export-subscriptions-slot',`${entity.name}Subscription`) #>
<#- slot('import-subscriptions-slot',`${entity.name}Subscription`) #>
import { ModelType, Type } from '../../../common';
import gql from 'graphql-tag';

export default new Type({
  type: ModelType.type,
  schema: gql`
    type #{entity.name}Subscription {
      mutation: MutationKind!
      node: #{entity.name}!
      payload: #{entity.name}SubscriptionPayload
      updatedFields: [String]
      <#-if(hasConnections){#>
      previous: Update#{entity.name}SubscriptionPayload
      <#-} else {#>
      previous: #{entity.name}SubscriptionPayload
      <#-}#>
    }
  `,
});

<#- if(hasConnections) {#>
<#-for ( let connection of entity.connections ) {#>
<#-chunkStart(`./subscriptions/${connection.name}ArgsSubscriptionPayload.ts`); -#>
<#- slot('export-subscriptions-slot',`${connection.name}ArgsSubscriptionPayload`) #>
<#- slot('import-subscriptions-slot',`${connection.name}ArgsSubscriptionPayload`) #>

import { ModelType, Type } from '../../../common';
import gql from 'graphql-tag';

export default new Type({
  type: ModelType.type,
  schema: gql`
    type #{connection.name}ArgsSubscriptionPayload {
      #{entity.ownerFieldName}:ID!
      #{connection.refFieldName}:ID!
    <#- if (connection.fields.length > 0){#>
      #additional Edge fields
    <# connection.fields.forEach(f=>{-#>
      #{f.name}: #{f.type}
    <# });-#>
    <#-}#>
    }
  `,
});

<#-chunkStart(`./subscriptions/${connection.name}SubscriptionPayload.ts`); -#>
<#- slot('export-subscriptions-slot',`${connection.name}SubscriptionPayload`) #>
<#- slot('import-subscriptions-slot',`${connection.name}SubscriptionPayload`) #>
import { ModelType, Type } from '../../../common';
import gql from 'graphql-tag';

export default new Type({
  type: ModelType.type,
  schema: gql`
    type #{connection.name}SubscriptionPayload {
      args:#{connection.name}ArgsSubscriptionPayload
      relation: String
    }
  `,
});
<# }-#>

<#-chunkStart(`./subscriptions/${entity.name}SubscriptionPayload.ts`); -#>
<#- slot('export-subscriptions-slot',`${entity.name}SubscriptionPayload`) #>
<#- slot('import-subscriptions-slot',`${entity.name}SubscriptionPayload`) #>
import { ModelType, Type } from '../../../common';
import gql from 'graphql-tag';

export default new Type({
  type: ModelType.type,
  schema: gql`
    union #{entity.name}SubscriptionPayload = Update#{entity.name}SubscriptionPayload | <# entity.connections.forEach(function(item, index){-#>
    <#- if(index > 0){#> | <#} -#>
    #{item.name}SubscriptionPayload
    <#- })-#>
  `,
    resolver : {
    __resolveType(obj, context, info) {
      if (
  <#-entity.unionCheck.forEach((fname, index)=>{-#><#-if(index > 0){#> || <#}-#>obj.#{fname}
  <#-})-#>) {
        return "Update#{entity.name}SubscriptionPayload";
      }
  <#-for ( let connection of entity.connections ) {#>
      if (obj.args && obj.args.#{entity.ownerFieldName} && obj.args.#{connection.refFieldName}) {
        return "#{connection.name}SubscriptionPayload";
      }
  <#-}#>
      return null;
    }
  },
});

<#}-#>