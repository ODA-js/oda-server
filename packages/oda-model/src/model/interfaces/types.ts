export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  | 'metadata'
  | 'model'
  | 'query'
  | 'mutation'
  | 'package'
  | 'entity'
  | 'scalar'
  | 'mixin'
  | 'union'
  | 'enum'
  | 'enum-item'
  | 'field-base'
  | 'field'
  | 'enum-field'
  | 'entity-field'
  | 'relation-field'
  | 'relation'
  | 'operation'
  | 'ref'
  | RelationType;

export type Multiplicity = 'one' | 'many';
/**
 * kind of complex type
 */
export type ComplexTypeKind = 'enum' | 'entity';

export type OperationKind = 'create' | 'read' | 'update' | 'delete';

export type DirectiveLocation =
  | 'QUERY'
  | 'MUTATION'
  | 'SUBSCRIPTION'
  | 'FIELD'
  | 'FRAGMENT_DEFINITION'
  | 'FRAGMENT_SPREAD'
  | 'INLINE_FRAGMENT'
  | 'SCHEMA'
  | 'SCALAR'
  | 'OBJECT'
  | 'FIELD_DEFINITION'
  | 'ARGUMENT_DEFINITION'
  | 'INTERFACE'
  | 'UNION'
  | 'ENUM'
  | 'ENUM_VALUE'
  | 'INPUT_OBJECT'
  | 'INPUT_FIELD_DEFINITION';

export interface INamed {
  readonly name: string;
}

export interface FieldArgsForHash {
  type?: string;
  required?: boolean;
  defaultValue?: string;
}

export interface FieldArgs extends INamed, FieldArgsForHash {}
/**
 * Complext type can refer to existing type system
 */
export type EntityType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'entity';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type EnumType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'enum';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type FieldType = string | EnumType | EntityType;

export interface AsHash<T extends INamed> {
  [name: string]: T;
}

export function HashToMap<T extends INamed>(input: AsHash<T>): Map<string, T> {
  const res = Object.keys(input).map(name => {
    return [name, input[name] as T] as [string, T];
  });
  return new Map(res);
}

export function MapToHash<T extends INamed>(input: Map<string, T>): AsHash<T> {
  return [...input.entries()].reduce(
    (hash, [name, value]) => {
      hash[name] = value;
      return hash;
    },
    {} as AsHash<T>,
  );
}
