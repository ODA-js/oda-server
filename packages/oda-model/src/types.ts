import { readFileSync, writeFileSync } from 'fs';
import { IUpdatableBase } from './element';
import { get, set, has } from 'lodash';

export type Nullable<T> = { [P in keyof T]: T[P] | undefined | null };

/**
 *
 */
export type assignInput<S, I, V = any> = {
  /**
   * source object
   */
  src: S;
  /**
   * dest object
   */
  input: Nullable<I>;
  /**
   * field from source object which is assigned
   */
  field?: keyof S | keyof S & keyof I | string;
  /**
   * function that will call to assign value
   */
  effect?: (src: S, value: V) => void;
  /**
   * function that check if effect is allowed
   */
  allowEffect?: (src: S, value: V) => boolean;
  /**
   * is field required, so it can't be reset by set null.
   * in this case assign will throw
   */
  required?: boolean;
  /**
   * field from dest object which value will be assigned to dest field
   */
  inputField?: keyof I;
  /**
   * function that will call to set default value
   */
  setDefault?: (src: S) => void;
};

/**
 * check specified value from input and assign
 * @param options assign Options
 */
export function assignValue<S extends object, I extends object, V = any>(
  options: assignInput<S, I, V>,
) {
  let {
    src,
    input,
    field,
    effect,
    required,
    inputField,
    setDefault,
    allowEffect,
  } = options;

  if (!inputField) {
    inputField = field as keyof I;
  }
  if (has(input, inputField)) {
    // update value
    if (get(input, inputField) != null) {
      if (effect) {
        if (
          (allowEffect && allowEffect(src, get(input, inputField) as any)) ||
          !allowEffect
        ) {
          effect(src, get(input, inputField) as any);
        } else if (
          allowEffect &&
          !allowEffect(src, get(input, inputField) as any) &&
          setDefault instanceof Function
        ) {
          setDefault(src);
        }
      } else if (field) {
        set(src, field, get(input, inputField));
      }
    } else if (inputField && get(input, inputField) === null && required) {
      if (setDefault instanceof Function) {
        setDefault(src);
      } else if (field) {
        throw new Error("can't reset required value");
      }
    } else if (inputField && get(input, inputField) === null) {
      if (setDefault instanceof Function) {
        setDefault(src);
      } else if (field) {
        set(src, field, undefined);
      }
    }
  } else {
    // set default value
    if (field && get(src, field) == null) {
      if (setDefault instanceof Function) {
        setDefault(src);
      }
    }
  }
}

export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  | 'element'
  | 'metadata'
  | 'package-base'
  | 'model'
  | 'package'
  | 'entity'
  | 'mixin'
  | 'query'
  | 'mutation'
  | 'scalar'
  | 'union'
  | 'directive'
  | 'enum'
  | 'enum-item'
  | 'entity-base'
  | 'field-base'
  | 'simple-field'
  | 'enum-field'
  | 'entity-field'
  | 'relation-field'
  | 'relation'
  | 'relation-base'
  | 'operation'
  | 'ref'
  | 'args'
  | 'model-hook'
  | RelationType;

export type Multiplicity = 'one' | 'many';
/**
 * kind of complex type
 */
export type ComplexTypeKind = 'enum' | 'entity';

export type OperationKind =
  | 'create'
  | 'readOne'
  | 'readMany'
  | 'update'
  | 'delete'
  | 'addTo'
  | 'removeFrom';

export type DirectiveLocation =
  | 'QUERY'
  | 'MUTATION'
  | 'SUBSCRIPTION'
  | 'FIELD'
  | 'FRAGMENT_DEFINITION'
  | 'FRAGMENT_SPREAD'
  | 'INLINE_FRAGMENT'
  | 'SCHEMA'
  | 'SCALAR'
  | 'OBJECT'
  | 'FIELD_DEFINITION'
  | 'ARGUMENT_DEFINITION'
  | 'INTERFACE'
  | 'UNION'
  | 'ENUM'
  | 'ENUM_VALUE'
  | 'INPUT_OBJECT'
  | 'INPUT_FIELD_DEFINITION';

export interface INamed {
  readonly name: string;
}

/**
 * Complext type can refer to existing type system
 */
export type EntityType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'entity';
  /**
   * multiplicity of it
   */
  multiplicity?: Multiplicity;
};

export type EnumType = {
  /**
   * complex type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'enum';
  /**
   * multiplicity of it
   */
  multiplicity?: Multiplicity;
};

export type FieldType = string | EnumType | EntityType;

export interface AsHash<T extends Partial<INamed> & object> {
  [name: string]: T;
}

export type NamedArray<T extends INamed> = Array<T>;

function mergeDupes<O extends Partial<INamed> & object>(
  src: Map<string, O>,
  items: [string, O][],
  killDupes?: (obj: O, src: O) => void,
) {
  items.forEach(item => {
    if (killDupes) {
      const dupe = src.get(item[0]);
      if (dupe) {
        killDupes(dupe, item[1]);
      } else {
        src.set(item[0], item[1]);
      }
    } else {
      src.set(item[0], item[1]);
    }
  });
}

export function ArrayToMap<T extends INamed, O extends INamed = T>(
  input: NamedArray<T>,
  process?: (v: T, index: number) => O,
  killDupes?: (obj: O, src: O) => void,
) {
  const res = new Map<string, O>();
  mergeDupes(
    res,
    input.map((f, index) => {
      let result: O = (f as unknown) as O;
      if (process) {
        result = (process(f, index) as unknown) as O;
      }
      return [result.name, result] as [string, O];
    }),
    killDupes,
  );
  return res;
}

export function HashToMap<
  T extends Partial<INamed> & object,
  O extends Partial<INamed> & object = T
>(
  input: AsHash<T>,
  process?: (name: string, v: T) => O,
  killDupes?: (obj: O, src: O) => void,
): Map<string, O> {
  const res = new Map<string, O>();
  mergeDupes(
    res,
    Object.keys(input).map(key => {
      let result: O | T = (input[key] as unknown) as O;
      if (process) {
        result = process(key, (result as unknown) as T);
      } else {
        result = {
          name: key,
          ...result,
        };
      }
      return [result.name, result] as [string, O];
    }),
    killDupes,
  );
  return res;
}

export function HashToArray<T extends Partial<INamed> & object>(
  input: AsHash<T>,
  process?: (name: string, v: T) => T,
  killDupes?: (obj: T, src: T) => void,
): T[] {
  const res = new Map<string, T>();
  mergeDupes(
    res,
    Object.keys(input).map(key => {
      let result = input[key];
      if (process) {
        result = process(key, result);
      } else {
        result = {
          ...result,
          name: key,
        };
      }
      return [result.name, result] as [string, T];
    }),
    killDupes,
  );
  return [...res.values()];
}

export function ArrayToHash<T extends INamed>(
  input: Array<T>,
  process?: (v: T) => T,
  killDupes?: (obj: T, src: T) => void,
): AsHash<T> {
  const res = new Map<string, T>();
  mergeDupes(
    res,
    input.map(f => {
      let result: T = (f as unknown) as T;
      if (process) {
        result = (process(f) as unknown) as T;
      }
      return [result.name, result] as [string, T];
    }),
    killDupes,
  );
  return [...res.values()].reduce(
    (result, item) => {
      result[item.name] = item;
      return result;
    },
    {} as AsHash<T>,
  );
}

export function MapToHash<T extends INamed, O extends INamed = T>(
  input: Map<string, T>,
  process?: (name: string, v: T) => O,
): AsHash<O> {
  return [...input.entries()].reduce(
    (hash, [name, value]) => {
      hash[name] = process
        ? (process(name, value) as O)
        : ((value as unknown) as O);
      return hash;
    },
    {} as AsHash<O>,
  );
}

export function MapToArray<T extends INamed, O extends INamed = T>(
  input: Map<string, T>,
  process?: (name: string, v: T) => O,
): NamedArray<O> {
  return [...input.entries()].reduce(
    (r, curr) => {
      r.push(
        process
          ? (process(curr[0], curr[1]) as O)
          : ((curr[1] as unknown) as O),
      );
      return r;
    },
    [] as O[],
  );
}

export type Merger<I extends INamed> = (
  objValue: any,
  srcValue: any,
  key: keyof I,
  object: I,
  source: I,
  stack: string[],
) => any;
/**
 * it checks and create item
 * @param srcMap external source map
 * @param create constructor for creating new item
 */
export function AssignAndKillDupes<
  T extends IUpdatableBase & INamed,
  I extends INamed
>(srcMap: Map<string, T> | undefined, create: new (input: I) => T) {
  return (i: I | string) => {
    let res: T | undefined;
    /** if item is string */
    if (typeof i === 'string') {
      /** check if we have one in external map */
      res = srcMap && srcMap.get(i);
      if (!res) {
        /** if not-> create it an push to external map */
        res = new create({ name: i } as any);
        srcMap && srcMap.set(res.name, res);
      }
    } else {
      /** if item is input object */
      res = new create(i);
      /** if we have the same in external map */
      const original = srcMap && srcMap.get(res.name);
      if (original) {
        /** if we have have then merge it with this one*/
        original.mergeWith(res.toObject());
      } else {
        /** set as original one */
        srcMap && srcMap.set(res.name, res);
      }
    }
    return res;
  };
}

export function loadFromFile<T, I>(
  filename: string,
  create: new (inp: I) => T,
) {
  const txt = readFileSync(filename);
  return new create(JSON.parse(txt.toString()) as I);
}

export function saveToFile<T extends { toObject: () => any }>(
  filename: string,
  inp: T,
) {
  writeFileSync(filename, inp.toObject());
}
