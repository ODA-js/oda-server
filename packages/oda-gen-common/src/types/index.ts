import { StateOfFieldType } from './stateOfField';
import { StateOfConectionType } from './stateOfConnection';
import { PageInfoType } from './pageInfo';
import { DateType } from './date';
import { IdType } from './id';
import { ImageSizeType } from './imageSize';
import { JSONType } from './json';
import { GQLModule } from './empty';
import { MutationKindType } from './mutationType';
import { WhereBoolean } from './whereBoolean';
import { WhereDate } from './whereDate';
import { WhereFloat } from './whereFloat';
import { WhereID } from './whereID';
import { WhereInt } from './whereInt';
import { WhereString } from './whereString';
import { WhereListOfStrings } from './whereListOfStrings';
import { WhereMutationKind } from './whereMutationKind';
import { WhereJSON } from './whereJSON';

export {
  GQLModule,
  StateOfFieldType,
  StateOfConectionType,
  PageInfoType,
  DateType,
  IdType,
  ImageSizeType,
  JSONType,
  WhereBoolean,
  WhereDate,
  WhereFloat,
  WhereID,
  WhereInt,
  WhereString,
  WhereListOfStrings,
  WhereMutationKind,
  WhereJSON,
};

export class DefaultTypes extends GQLModule {
  protected _name = 'DefaultTypes';
  protected _composite = [
    new StateOfFieldType({}),
    new StateOfConectionType({}),
    new PageInfoType({}),
    new DateType({}),
    new IdType({}),
    new ImageSizeType({}),
    new JSONType({}),
    new MutationKindType({}),
    new WhereBoolean({}),
    new WhereDate({}),
    new WhereFloat({}),
    new WhereID({}),
    new WhereInt({}),
    new WhereString({}),
    new WhereListOfStrings({}),
    new WhereMutationKind({}),
    new WhereJSON({}),
  ];
}
