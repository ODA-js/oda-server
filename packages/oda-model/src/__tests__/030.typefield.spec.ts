import 'jest';
import { TypeField } from '../typefield';
import { EnumType } from '../types';

describe('TypeField', () => {
  it('default', () => {
    const res = new TypeField({
      name: 'A',
    });
    expect(res.name).toBe('A');
    expect(res.type).toBe('string');
    expect(res.modelType).toBe('input-simple-field');
    expect(res.multiplicity).toBe('one');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });

  it('default value only for persistent fields', () => {
    const res = new TypeField({
      name: 'A',
      defaultValue: 'NAME_OF_DEFAULT_VALUE',
    });
    expect(res.name).toBe('A');
    expect(res.defaultValue).toBe('NAME_OF_DEFAULT_VALUE');
  });

  it('required by default is false', () => {
    const res = new TypeField({
      name: 'A',
    });
    expect(res.name).toBe('A');
    expect(res.required).toBeFalsy();
  });
  it('required is setup', () => {
    const res = new TypeField({
      name: 'A',
      required: true,
    });
    expect(res.name).toBe('A');
    expect(res.required).toBeTruthy();
    expect(res.toObject()).toMatchSnapshot('required');
  });

  it('default enum', () => {
    const res = new TypeField({
      name: 'A',
      type: {
        type: 'enum',
        multiplicity: 'many',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('input-enum-field');
    expect(res.multiplicity).toBe('many');
    expect((res.type as EnumType).type).toBe('enum');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });

  it('default enum list', () => {
    const res = new TypeField({
      name: 'A',
      multiplicity: 'many',
      type: {
        type: 'enum',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('input-enum-field');
    expect(res.multiplicity).toBe('many');
    expect((res.type as EnumType).type).toBe('enum');
    expect((res.type as EnumType).multiplicity).toBe('many');
  });
  it('default enum multiplicity == one', () => {
    const res = new TypeField({
      name: 'A',
      type: {
        type: 'enum',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('input-enum-field');
    expect(res.multiplicity).toBe('one');
    expect((res.type as EnumType).type).toBe('enum');
    expect((res.type as EnumType).multiplicity).toBe('one');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
