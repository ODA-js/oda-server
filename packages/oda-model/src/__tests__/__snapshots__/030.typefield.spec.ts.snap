// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeField default enum multiplicity == one: default 1`] = `
TypeField {
  Symbol(internal): Object {
    "multiplicity": "one",
    "name": "A",
    "type": Object {
      "multiplicity": "one",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {},
}
`;

exports[`TypeField default enum multiplicity == one: toObject 1`] = `
Object {
  "defaultValue": undefined,
  "description": undefined,
  "metadata": Object {},
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": Object {
    "multiplicity": "one",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`TypeField default enum: default 1`] = `
TypeField {
  Symbol(internal): Object {
    "multiplicity": "many",
    "name": "A",
    "type": Object {
      "multiplicity": "many",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {},
}
`;

exports[`TypeField default enum: toObject 1`] = `
Object {
  "defaultValue": undefined,
  "description": undefined,
  "metadata": Object {},
  "multiplicity": "many",
  "name": "A",
  "title": undefined,
  "type": Object {
    "multiplicity": "many",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`TypeField default: default 1`] = `
TypeField {
  Symbol(internal): Object {
    "multiplicity": "one",
    "name": "A",
    "type": "string",
  },
  Symbol(metadata): Object {},
}
`;

exports[`TypeField default: toObject 1`] = `
Object {
  "defaultValue": undefined,
  "description": undefined,
  "metadata": Object {},
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": "string",
}
`;

exports[`TypeField required is setup: required 1`] = `
Object {
  "defaultValue": undefined,
  "description": undefined,
  "metadata": Object {
    "required": true,
  },
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": "string",
}
`;
